codec-http2/src/main/java/io/netty/handler/codec/http2/HttpToHttp2ConnectionHandler.java

```diff
@@ -57,45 +65,64 @@ private int getStreamId(HttpHeaders httpHeaders) throws Exception {
     }
 
     /**
-     * Handles conversion of a {@link FullHttpMessage} to HTTP/2 frames.
+     * Handles conversion of {@link HttpMessage} and {@link HttpContent} to HTTP/2 frames.
      */
     @Override
     public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
-        if (msg instanceof FullHttpMessage) {
-            FullHttpMessage httpMsg = (FullHttpMessage) msg;
-            boolean hasData = httpMsg.content().isReadable();
-            boolean httpMsgNeedRelease = true;
-            SimpleChannelPromiseAggregator promiseAggregator = null;
-            try {
+
+        if (!(msg instanceof HttpMessage || msg instanceof HttpContent)) {
+            ctx.write(msg, promise);
+            return;
+        }
+
+        boolean release = true;
+        SimpleChannelPromiseAggregator promiseAggregator =
+                new SimpleChannelPromiseAggregator(promise, ctx.channel(), ctx.executor());
+        try {
+            Http2ConnectionEncoder encoder = encoder();
+            boolean endStream = false;
+            if (msg instanceof HttpMessage) {
+                final HttpMessage httpMsg = (HttpMessage) msg;
+
                 // Provide the user the opportunity to specify the streamId
-                int streamId = getStreamId(httpMsg.headers());
+                currentStreamId = getStreamId(httpMsg.headers());
 
                 // Convert and write the headers.
                 Http2Headers http2Headers = HttpUtil.toHttp2Headers(httpMsg);
-                Http2ConnectionEncoder encoder = encoder();
-
-                if (hasData) {
-                    promiseAggregator = new SimpleChannelPromiseAggregator(promise, ctx.channel(), ctx.executor());
-                    encoder.writeHeaders(ctx, streamId, http2Headers, 0, false, promiseAggregator.newPromise());
-                    httpMsgNeedRelease = false;
-                    encoder.writeData(ctx, streamId, httpMsg.content(), 0, true, promiseAggregator.newPromise());
-                    promiseAggregator.doneAllocatingPromises();
-                } else {
-                    encoder.writeHeaders(ctx, streamId, http2Headers, 0, true, promise);
-                }
-            } catch (Throwable t) {
-                if (promiseAggregator == null) {
-                    promise.tryFailure(t);
-                } else {
-                    promiseAggregator.setFailure(t);
+                endStream = msg instanceof FullHttpMessage && !((FullHttpMessage) msg).content().isReadable();
```

Comment:
As far as I'm aware, yes.  If the FullHttpMessage don't not contain any data, the only HTTP/2 frame to write is the HEADERS frame.  You can see the old code did the same thing.